File: src/main/java/com/wxl/dyttcrawler/scheduler/BatchDuplicateRemovedScheduler.java
package com.wxl.dyttcrawler.scheduler;

import com.wxl.dyttcrawler.scheduler.local.BatchHashSetDuplicateRemover;
import org.apache.commons.collections.CollectionUtils;
import us.codecraft.webmagic.Request;
import us.codecraft.webmagic.Task;
import us.codecraft.webmagic.scheduler.DuplicateRemovedScheduler;
import us.codecraft.webmagic.scheduler.component.DuplicateRemover;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Create by wuxingle on 2020/7/20
 * 批量移除重复request的scheduler
 */
public abstract class BatchDuplicateRemovedScheduler
        extends DuplicateRemovedScheduler
        implements BatchScheduler {

    public BatchDuplicateRemovedScheduler() {
        setDuplicateRemover(new BatchHashSetDuplicateRemover());
    }

    @Override
    public void push(List<Request> requests, Task task) {
        List<Request> shouldPutRequests = new ArrayList<>();
        List<Request> shouldFilterRequests = new ArrayList<>();
        for (Request request : requests) {
            if (shouldReserved(request) || noNeedToRemoveDuplicate(request)) {
                shouldPutRequests.add(request);
            } else {
                shouldFilterRequests.add(request);
            }
        }
        if (CollectionUtils.isNotEmpty(shouldFilterRequests)) {
            shouldFilterRequests = getDuplicateRemover().filterDuplicate(shouldFilterRequests, task);
            shouldPutRequests.addAll(shouldFilterRequests);
        }

        if (CollectionUtils.isNotEmpty(shouldPutRequests)) {
            if (logger.isDebugEnabled()) {
                logger.debug("push to queue {}", shouldPutRequests.stream()
                        .map(Request::getUrl).collect(Collectors.toList()));
            }
            pushWhenNoDuplicate(shouldPutRequests, task);
        }
    }

    @Override
    public BatchDuplicateRemover getDuplicateRemover() {
        return (BatchDuplicateRemover) super.getDuplicateRemover();
    }

    @Override
    public BatchDuplicateRemovedScheduler setDuplicateRemover(DuplicateRemover duplicatedRemover) {
        if (!(duplicatedRemover instanceof BatchDuplicateRemover)) {
            throw new IllegalStateException("DuplicateRemover must is batch impl");
        }
        super.setDuplicateRemover(duplicatedRemover);
        return this;
    }

    protected abstract void pushWhenNoDuplicate(Request request, Task task);

    protected void pushWhenNoDuplicate(Collection<Request> requests, Task task) {
        for (Request request : requests) {
            pushWhenNoDuplicate(request, task);
        }
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/scheduler/BatchDuplicateRemovedScheduler.kt
package com.wxl.dyttcrawler.scheduler

import com.wxl.dyttcrawler.scheduler.local.BatchHashSetDuplicateRemover
import us.codecraft.webmagic.Request
import us.codecraft.webmagic.Task
import us.codecraft.webmagic.scheduler.DuplicateRemovedScheduler
import us.codecraft.webmagic.scheduler.component.DuplicateRemover

/**
 * Create by wuxingle on 2021/10/11
 * 批量移除重复request的scheduler
 */
abstract class BatchDuplicateRemovedScheduler : DuplicateRemovedScheduler(), BatchScheduler {

    init {
        duplicateRemover = BatchHashSetDuplicateRemover()
    }

    override fun push(requests: List<Request>, task: Task) {
        val shouldPutRequests = mutableListOf<Request>()
        val shouldFilterRequests = mutableListOf<Request>()

        for (request in requests) {
            if (shouldReserved(request) || noNeedToRemoveDuplicate(request)) {
                shouldPutRequests.add(request)
            } else {
                shouldFilterRequests.add(request)
            }
        }
        if (shouldFilterRequests.isNotEmpty()) {
            shouldPutRequests.addAll(duplicateRemover.filterDuplicate(shouldFilterRequests, task))
        }

        if (shouldPutRequests.isNotEmpty()) {
            if (logger.isDebugEnabled) {
                logger.debug("push to queue {}", shouldPutRequests.map(Request::getUrl))
            }
            pushWhenNoDuplicate(shouldPutRequests, task)
        }
    }

    override fun getDuplicateRemover(): BatchDuplicateRemover =
        super.getDuplicateRemover() as BatchDuplicateRemover


    final override fun setDuplicateRemover(duplicatedRemover: DuplicateRemover): BatchDuplicateRemovedScheduler {
        if (duplicatedRemover !is BatchDuplicateRemover) {
            throw IllegalStateException("DuplicateRemover must is batch impl")
        }
        super.setDuplicateRemover(duplicatedRemover)
        return this
    }

    abstract override fun pushWhenNoDuplicate(request: Request, task: Task)

    open fun pushWhenNoDuplicate(requests: Collection<Request>, task: Task) {
        requests.forEach { pushWhenNoDuplicate(it, task) }
    }
}
